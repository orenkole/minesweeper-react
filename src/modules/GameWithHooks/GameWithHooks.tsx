import React, {FC, useMemo, useState} from "react";

import { GameArea } from "@/components/Game/GameArea";
import { CellState, Coords, emptyFieldGenerator, Field, fieldGenerator } from "@/helpers/Field";
import { ScoreBoard } from "@/components/ScoreBoard";
import { Wrapper } from "@/components/Game/Wrapper";
import { GameOver } from "@/components/Game/GameOver";
import { Grid } from "@/components/Grid/Grid";
import { Top } from "@/components/Top/Top";
import { GameLevels, GameSettings, LevelNames } from "../GameSettings";
import { openCell } from "@/helpers/openCell";

export const GameWithHooks: FC = () => {
	const [level, setLevel] = useState<LevelNames>('beginner');

	const [size, bombs] = GameSettings[level];

	const [gameField, setGameField] = useState<Field>(
		fieldGenerator(size, bombs / (size * size))
	);

	const [playerField, setPlayerField] = useState<Field>(
		emptyFieldGenerator(size, CellState.hidden)
	)

	const onClick = (coords: Coords) => {
		const newPlayerField = openCell(coords, playerField, gameField);
		setPlayerField([...newPlayerField]);
	}

	const onResetHandler = ([size, bombs]: [number, number]) => {
		const newGameField = fieldGenerator(size, bombs / (size * size))
		const newPlayerField = emptyFieldGenerator(size, CellState.hidden);
		setGameField([...newGameField])
		setPlayerField([...newPlayerField])
	}

	const onChangeLevel = ({
		target: {value: level},
	}: React.ChangeEvent<HTMLSelectElement>) => {
		setLevel(level as LevelNames);
		const newSettings = GameSettings[level];
		onResetHandler(newSettings)
	}

	const onReset = () => onResetHandler([size, bombs]);

	return (
		<Wrapper>
			<Top feature="Flag" firstAction="right click">Minesweeper</Top>
			<GameArea>
				<ScoreBoard
					time="0"
					mines="10"
					levels={GameLevels}
					onReset={onReset}
					onChange={onChangeLevel}
					defaultLevel={level}
				/>
				<GameOver onClick={() => null} isWin={true} />
				<Grid
					onClick={onClick}
					onContextMenu={() => null}
				>
					{playerField}
				</Grid>
			</GameArea>
		</Wrapper>
	)
};