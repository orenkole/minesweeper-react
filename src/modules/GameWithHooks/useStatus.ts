import {useState} from "react";

export interface ReturnType {
	isGameOver: boolean;
	isWin: boolean;
	isGameStart: boolean;
	setNewGame: () => void;
	setInProgress: () => void;
	setGameWin: () => void;
	setGameLoose: () => void;
}

export enum GameStatuses {
	NewGame,
	InProgress,
	Win,
	Loose,
}

export const useStatus = (): ReturnType => {
	const {NewGame, InProgress, Win, Loose} = GameStatuses;

	const [isGameOver, setIsGameOver] = useState(false);
	const [isWin, setIsWin] = useState(false);
	const [isGameStart, setIsGameStart] = useState(false);

	const setGameStatus = (status: GameStatuses) => {
		setIsGameStart(status === InProgress);
		setIsWin(status === Win);
		setIsGameOver([Win, Loose].includes(status));
	}

	const setNewGame = () => setGameStatus(NewGame);
	const setInProgress = () => setGameStatus(InProgress);
	const setGameWin = () => setGameStatus(Win);
	const setGameLoose = () => setGameStatus(Loose);


	return {
		isGameOver,
		isWin,
		isGameStart,
		setNewGame,
		setInProgress,
		setGameWin,
		setGameLoose,
	}
}