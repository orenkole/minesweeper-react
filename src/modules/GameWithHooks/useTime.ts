import {useState, useEffect} from "react";

export const useTime = (isGameStart: boolean, isGameOver: boolean): [number, () => void ]  => {
	const [time, setTime] = useState(0);
	
	useEffect(() => {
		let interval: NodeJS.Timeout;
		if(isGameStart) {
			interval = setInterval(() => {setTime(time + 1)}, 1000);
			if(isGameOver) {
				clearInterval(interval);
			}
		}
		return () => {clearInterval(interval)}
	},
	// Stryker disable next-line ArrayDeclaration
	[isGameOver, isGameStart, time])

	const onReset = () => setTime(0);

	return [time, onReset];
}