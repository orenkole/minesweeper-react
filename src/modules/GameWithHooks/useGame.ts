import React, {useState} from "react";

import { CellState, Coords, emptyFieldGenerator, Field, fieldGenerator } from "@/core/Field";
import { GameSettings, LevelNames } from "../GameSettings";
import { openCell } from "@/core/openCell";
import { setFlag } from "@/core/setFlag";
import { useTime } from "./useTime";


interface ReturnType {
	level: LevelNames;
	isGameOver: boolean;
	isWin: boolean;
	settings: [number, number];
	playerField: Field;
	gameField: Field;
	onClick: (coords: Coords) => void;
	onContextMenu: (coords: Coords) => void;
	onChangeLevel: ({target: {value: level}}: {target: {value: LevelNames}}) => void;
	onReset: () => void;
	time: number;
	flagCounter: number;
	isGameStart : boolean;
}

export const useGame = (): ReturnType => {
	const [level, setLevel] = useState<LevelNames>('beginner');

	const [isGameOver, setIsGameOver] = useState(false);

	const [isWin, setIsWin] = useState(false); 

	const [isGameStart, setIsGameStart] = useState(false);

	const [time, resetTime] = useTime(isGameStart, isGameOver);

	const [flagCounter, setFlagCounter] = useState(0);

	const setGameOver = (isSolved = false) => {
		setIsGameOver(true);
		setIsWin(isSolved);
	}

	const [size, bombs] = GameSettings[level];

	const [gameField, setGameField] = useState<Field>(
		fieldGenerator(size, bombs / (size * size))
	);

	const [playerField, setPlayerField] = useState<Field>(
		emptyFieldGenerator(size, CellState.hidden)
	)

	const onClick = (coords: Coords) => {
		!isGameStart && setIsGameStart(true);
		try {
			const [newPlayerField, isSolved] = openCell(coords, playerField, gameField);
			if(isSolved) {
				setGameOver(isSolved);
			}
			setPlayerField([...newPlayerField]);
		} catch (err) {
			// on loose we reveal all field
			setPlayerField([...gameField])
			setGameOver(false);
		}
	}

	const onContextMenu = (coords: Coords) => {
		!isGameStart && setIsGameStart(true);
		const [newPlayerField, isSolved, newFlagCounter] = 
			setFlag(coords, playerField, gameField, flagCounter, bombs);
		setFlagCounter(newFlagCounter)
		if(isSolved) {
			setGameOver(isSolved);
		}
		setPlayerField([...newPlayerField])
	}

	const onResetHandler = ([size, bombs]: [number, number]) => {
		const newGameField = fieldGenerator(size, bombs / (size * size))
		const newPlayerField = emptyFieldGenerator(size, CellState.hidden);
		setGameField([...newGameField])
		setPlayerField([...newPlayerField])
		setIsGameOver(false);
		setIsWin(false);
		setIsGameStart(false);
		resetTime();
	}

	const onChangeLevel = ({
		target: {value: level},
	}: {target: {value: LevelNames}}) => {
		setLevel(level as LevelNames);
		const newSettings = GameSettings[level];
		onResetHandler(newSettings)
	}

	const onReset = () => onResetHandler([size, bombs]);

	return {
		level,
		isGameOver,
		isWin,
		settings: [size, bombs],
		playerField,
		gameField,
		onClick,
		onContextMenu,
		onChangeLevel,
		onReset,
		time,
		flagCounter,
		isGameStart,
	}
}