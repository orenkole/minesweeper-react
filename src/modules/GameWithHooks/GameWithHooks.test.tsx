import React from "react";
import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { GameWithHooks } from "./GameWithHooks";

const mockOnClick = jest.fn();
const mockOnChangeLevel = jest.fn();
const mockOnReset = jest.fn();

jest.mock('./useGame', () => ({
	__esModule: true,
	useGame: () => ({
		level: 'beginner',
		isGameOver: 'beginner',
		isWin: false,
		settings: [9, 10],
		playerField: [
			[10, 10],
			[10, 10],
		],
		onClick: mockOnClick,
		onChangeLevel: mockOnChangeLevel,
		onReset: mockOnReset,
	})
}))

beforeEach(() => {
	jest.clearAllMocks()
})

describe("GameWithHooks test cases", () => {
	it("Render game field by default", () => {
		const {asFragment} = render(<GameWithHooks />)
		expect(asFragment).toMatchSnapshot();
	})
	it("Cell click works fine", () => {
		render(<GameWithHooks />)
		userEvent.click(screen.getByTestId('0,0'))
		expect(mockOnClick).toHaveBeenCalled();
	})
	it("Reset handler works fine", () => {
		render(<GameWithHooks />)
		userEvent.click(screen.getByRole('button'))
		expect(mockOnReset).toHaveBeenCalled();
	})
	it("Change level works fine", () => {
		render(<GameWithHooks />)
		userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
		expect(mockOnChangeLevel).toHaveBeenCalled();
	})
	it("Game over works fine", () => {
		render(<GameWithHooks />)
		userEvent.click(screen.getByText('🙁'));
		expect(mockOnReset).toHaveBeenCalled();
	})
})