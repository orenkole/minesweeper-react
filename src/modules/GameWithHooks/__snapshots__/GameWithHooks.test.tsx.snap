// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render game field by default 1`] = `[Function]`;
